File: proyecto_en_java/src/NormalizedUtility.java
import java.lang.Math;

public class NormalizedUtility {
    // Ana
    //Atributos de la Clase
    private String appName;
    private long reviewId;             //Id de la reseña
    private long timestampCreated;     //Fecha/hora en que se creó la reseña
    private boolean recommended;       //Si el usuario recomienda el juego (TRUE/FALSE)
    private int votesHelpful;          //Número de votos útiles recibidos
    private double weightedVoteScore;  //Puntaje de voto ponderado
    private double utilidadNormalizada;//Resultado del cálculo de utilidad normalizada

    public NormalizedUtility(String appName, long reviewId, long timestampCreated,
                             boolean recommended, int votesHelpful) {
        this.appName = appName;
        this.reviewId = reviewId;
        this.timestampCreated = timestampCreated;
        this.recommended = recommended;
        this.votesHelpful = votesHelpful;
        this.weightedVoteScore = 0.0; // Valor por defecto
        this.utilidadNormalizada = calcularUtilidadNormalizada();
    }

    // Getters
    public String getAppName() {
        return appName;
    }
    public long getReviewId() {
        return reviewId;
    }
    public long getTimestampCreated() {
        return timestampCreated;
    }
    public boolean isRecommended() {
        return recommended;
    }
    public int getVotesHelpful() {
        return votesHelpful;
    }
    public double getWeightedVoteScore() {
        return weightedVoteScore;
    }
    public double getUtilidadNormalizada() {
        return utilidadNormalizada;
    }

    //Calcula la utilidad normalizada basada en los atributos
    private double calcularUtilidadNormalizada() {
        double base = (recommended ? 1 : 0)
                + (votesHelpful / 10.0)
                + (weightedVoteScore * 2);
        return Math.min(base / 3.0, 1.0);
    }
}

File: algorithms/NormalizedUtility.py
import math
from typing import List

# CLASE: NormalizedUtility
class NormalizedUtility:
    # Ana
    # Atributos de la Clase
    
    def __init__(self, app_name: str, review_id: int, timestamp_created: int, 
                 recommended: bool, votes_helpful: int):
        
        self.app_name = app_name
        self.review_id = review_id
        self.timestamp_created = timestamp_created
        self.recommended = recommended
        self.votes_helpful = votes_helpful
        
        # Atributos extra que no se reciben por el constructor (se inicializan en 0.0)
        self.weighted_vote_score = 0.0
        self.utilidad_normalizada = self._calcular_utilidad_normalizada()

    def _calcular_utilidad_normalizada(self) -> float:
        # Calcula la utilidad normalizada.
        base = (1.0 if self.recommended else 0.0) \
               + (self.votes_helpful / 10.0) \
               + (self.weighted_vote_score * 2)
        return min(base / 3.0, 1.0)

    # Getters y __repr__ (simplificado para el código)
    def get_app_name(self) -> str: return self.app_name
    def get_review_id(self) -> int: return self.review_id
    def get_timestamp_created(self) -> int: return self.timestamp_created
    def get_votes_helpful(self) -> int: return self.votes_helpful
    def get_utilidad_normalizada(self) -> float: return self.utilidad_normalizada

    def __repr__(self) -> str:
        # Imprime los datos
        return f"App: {self.app_name}, ID: {self.review_id}, Time: {self.timestamp_created}"